File: src/main/java/aurocosh/divinefavor/common/item/talismans/arrow/ArrowTalismanDisarm.java
package aurocosh.divinefavor.common.item.talismans.arrow;

import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowOptions;
import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowType;
import aurocosh.divinefavor.common.item.talismans.arrow.base.ItemArrowTalisman;
import aurocosh.divinefavor.common.spirit.base.ModSpirit;
import aurocosh.divinefavor.common.util.InventoryIndexes;
import aurocosh.divinefavor.common.util.SlotData;
import aurocosh.divinefavor.common.util.UtilPlayer;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.projectile.EntityArrow;
import net.minecraft.item.ItemStack;

import java.util.EnumSet;

public class ArrowTalismanDisarm extends ItemArrowTalisman {
    public ArrowTalismanDisarm(String name, ModSpirit spirit, int favorCost, int color, double arrowDamage, EnumSet<ArrowOptions> options, ArrowType arrowType) {
        super(name, spirit, favorCost, color, arrowDamage, options, arrowType);
    }

    @Override
    protected void performActionServer(EntityLivingBase target, EntityLivingBase shooter, EntityArrow arrow) {
        if (target instanceof EntityPlayer) {
            EntityPlayer player = (EntityPlayer) target;
            moveStackToMainInventory(player, player.inventory.currentItem);
            moveStackToMainInventory(player, InventoryIndexes.Offhand.getValue());
        }
    }

    private void moveStackToMainInventory(EntityPlayer player, int slot) {
        ItemStack stack = player.inventory.getStackInSlot(slot);
        if (stack.isEmpty())
            return;
        SlotData slotData = UtilPlayer.findStackInMainInventory(player, ItemStack::isEmpty);
        if (slotData.isValid())
            UtilPlayer.swapStacks(player, slot, slotData.slotIndex);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/arrow_talismans/ArrowTalismanDisarm.kt
package aurocosh.divinefavor.common.item.arrow_talismans

import aurocosh.divinefavor.common.entity.projectile.EntitySpellArrow
import aurocosh.divinefavor.common.item.arrow_talismans.base.ArrowOptions
import aurocosh.divinefavor.common.item.arrow_talismans.base.ArrowType
import aurocosh.divinefavor.common.item.arrow_talismans.base.ItemArrowTalisman
import aurocosh.divinefavor.common.lib.enums.InventoryIndexes
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.util.UtilPlayer
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import java.awt.Color
import java.util.*

class ArrowTalismanDisarm(name: String, spirit: ModSpirit, favorCost: Int, color: Color, arrowDamage: Double, options: EnumSet<ArrowOptions>, arrowType: ArrowType) : ItemArrowTalisman(name, spirit, favorCost, color, arrowDamage, options, arrowType) {

    override fun performActionServer(target: EntityLivingBase?, shooter: EntityLivingBase, spellArrow: EntitySpellArrow, blockPos: BlockPos?, sideHit: EnumFacing?): Boolean {
        if (target is EntityPlayer) {
            moveStackToMainInventory(target, target.inventory.currentItem)
            moveStackToMainInventory(target, InventoryIndexes.Offhand.value)
        }
        return true
    }

    private fun moveStackToMainInventory(player: EntityPlayer, slot: Int) {
        val stack = player.inventory.getStackInSlot(slot)
        if (stack.isEmpty)
            return
        val slotData = UtilPlayer.findStackInMainInventory(player) { obj: ItemStack -> obj.isEmpty }
        if (slotData.isValid)
            UtilPlayer.swapStacks(player, slot, slotData.slotIndex)
    }
}
